File: java/leetcode/product-of-array-except-self.java
class Solution {
    static int[] productExceptSelf(int[] nums) {
        int[] b = new int[nums.length];
        b[0] = 1;
        int a = 1;
        for (int i = 0; i <= nums.length - 2; i++) {
            a *= nums[i];
            b[i + 1] = a;
        }
        a = 1;
        for (int i = (nums.length - 1); i >= 1; i--) {
            a *= nums[i];
            b[i - 1] *= a;
        }
        return b;
    }
}

File: python/leetcode/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        b = [0] * len(nums)
        b[0] = 1
        a = 1
        for i in range(len(nums) - 1):
            a *= nums[i]
            b[i + 1] = a
        a = 1
        for i in range(len(nums) - 1, 0, -1):
            a *= nums[i]
            b[i - 1] *= a
        return b
